firedrake/assemble.py

```diff
@@ -91,21 +92,104 @@ def assemble(expr, tensor=None, bcs=None, form_compiler_parameters=None,
         raise TypeError("Unknown keyword arguments '%s'" % ', '.join(kwargs.keys()))
 
     if isinstance(expr, (ufl.form.Form, slate.TensorBase)):
-        loops = _assemble(expr, tensor=tensor, bcs=solving._extract_bcs(bcs),
-                          form_compiler_parameters=form_compiler_parameters,
-                          mat_type=mat_type,
-                          sub_mat_type=sub_mat_type, appctx=appctx,
-                          diagonal=diagonal,
-                          options_prefix=options_prefix)
-        for l in loops:
-            m = l()
-        return m
+        mat_type, sub_mat_type = get_mat_type(mat_type, sub_mat_type, expr.arguments())
+        assembly_rank = _get_assembly_rank(expr, diagonal)
+        if assembly_rank == AssemblyRank.SCALAR:
+            return _assemble_scalar(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
+        elif assembly_rank == AssemblyRank.VECTOR:
+            return _assemble_vector(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, diagonal=diagonal, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
+        elif assembly_rank == AssemblyRank.MATRIX:
+            return _assemble_matrix(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
     elif isinstance(expr, ufl.core.expr.Expr):
         return assemble_expressions.assemble_expression(expr)
     else:
         raise TypeError(f"Unable to assemble: {expr}")
 
 
+def _assemble_scalar(expr, scalar, assemble_now, **kwargs):
+    if len(expr.arguments()) != 0:
+        raise ValueError("Can't assemble a 0-form with arguments")
+    if scalar:
+        raise ValueError("Can't assemble 0-form into existing tensor")
+
+    scalar = _make_scalar()
+    _assemble_expr(expr, scalar, diagonal=False, assemble_now=assemble_now, **kwargs)
+    return scalar.data[0]
+
+
+def _assemble_vector(expr, vector, *, diagonal, assemble_now, **kwargs):
```

Comment:
I did not realise that you could request _required_ keyword arguments like this!

Reply:
Neither did I. I just tried it and it worked!